### PR TITLE
feat(#19183): add namespace input to chronicle sink

### PR DESCRIPTION
**Pull request by ChocPanda**
feat(#19183): add namespace input to chronicle sink

### CODE REVIEW

**Review by hestonhoffman — APPROVED** [review]

**Comment by ChocPanda** [comment]
Apologies, I needed to amend the commit to the correct email address in order to sign the CLA.

**Review by StephenWakely — CHANGES_REQUESTED** [review]
I think it would make sense to make this field templatable so different events can be sent to different namespaces. There's a slight complexity in that we would then need to partition the event batches by namespace. If you are up for the challenge, I can walk you through what would be required.

If you don't have time, it would be ok to merge this as is, since it isn't introducing a regression.

We do need an entry for our changelog see [here](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md) for details.

Let me know if you want to make that additional change.

**Comment by ChocPanda** [comment]
Hey @StephenWakely, is there anything I need to do to get this merged?

**Comment by ChocPanda** [comment]
Hey @StephenWakely, Can we do both? Templatable namespacing sounds really interesting but I am in the process of deploying vector at the moment and I need this feature with a static namespace.

I'll add the documentation to the changelog this evening

**Comment by StephenWakely** [comment]
> Hey @StephenWakely, Can we do both? Templatable namespacing sounds really interesting but I am in the process of deploying vector at the moment and I need this feature with a static namespace.
> 
> I'll add the documentation to the changelog this evening

Yes. If I understand you correctly, a template field will only template the parts of the string delimited with `{{..}}`. So a field of `namespace` will render as `namespace`, but a field of `namespace-{{ foo }}` and an event of `{ "foo": 32 }` will render as `namespace-32`.

**Comment by StephenWakely** [comment]
@ChocPanda 

We have this CI error now:
> File "changelog.d/19183_chronicle_namespace_support.enhancement.md" contains trailing spaces in some of the lines

Our CI can be annoying sometimes, not gonna lie!

**Comment by ChocPanda** [comment]
Hey @StephenWakely , I think I resolved the issue, unfortunately I was able to run the check locally to confirm

**Comment by StephenWakely** [comment]
Hi, the integration test failure is not an issue and can be ignored. The [other failure](https://github.com/vectordotdev/vector/actions/runs/7987827217/job/21812455406?pr=19398) is because our documentation is now out of sync.

Can you run `make generate-component-docs` and check in the resulting changes? This should get everything back in sync.

**Comment by ChocPanda** [comment]
Hey @StephenWakely, fixed the docs error

File: src/sinks/gcp/chronicle_unstructured.rs [hunk 1]
```diff
@@ -127,6 +127,10 @@ pub struct ChronicleUnstructuredConfig {
     #[configurable(metadata(docs::examples = "c8c65bfa-5f2c-42d4-9189-64bb7b939f2c"))]
     pub customer_id: String,
 
+    /// User-configured environment namespace to identify the data domain the logs originated from.
+    #[configurable(metadata(docs::examples = "production"))]
+    pub namespace: String,
```
Comment (StephenWakely) [inline src/sinks/gcp/chronicle_unstructured.rs:132]: This should be optional so it doesn't affect users that are currently not configuring a namespace...

```suggestion
    pub namespace: Option<String>,
```

**Review by StephenWakely — CHANGES_REQUESTED** [review]
Hi, sorry I didn't pick up on this sooner. The `namespace` option should be optional so it isn't a breaking change for users that are currently not specifying one. 

The [docs](https://cloud.google.com/chronicle/docs/reference/ingestion-api#unstructuredlogentries) do imply that this field is required, but then don't show that field in the examples - implying that it isn't required. So it's a little bit confusing.

Would you be able to make this change?

**Comment by StephenWakely** [comment]
@ChocPanda We are getting some CI failures.

- `src/sinks/gcp/chronicle_unstructured.rs` needs to be run through `rustfmt` so it is properly formatted.

- The integration tests are failing with:

```
thread 'sinks::gcp::chronicle_unstructured::integration_tests::publish_events' panicked at src/sinks/gcp/chronicle_unstructured.rs:666:14:
Extracting pull data failed: reqwest::Error { kind: Decode, source: Error("missing field `namespace`", line: 1, column: 214) }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
test sinks::gcp::chronicle_unstructured::integration_tests::publish_events ... FAILED
```

The integration tests runs against a container that we set up to emulate Chronicle. It is likely that that error is caused because that server is not expecting the `namespace` field. 

https://github.com/vectordotdev/chronicle-emulator/blob/main/src/data.rs#L18

I will see if I can sort that out in a minute.

**Review by StephenWakely — APPROVED** [review]
Awesome. That seems to have done the trick. Thanks.

**Pull request merged**